### PR TITLE
fix: Use statefull analyzers in generator

### PR DESCRIPTION
**Pull request by SandPod**
Uses stateful analyzers to analyze both endpoints and model files in the generator.

This PR is a preparation for supporting endpoint and model files located anywhere in the project that is tracked here  https://github.com/serverpod/serverpod/issues/964

This PR also introduces additional filtering on the watched files to only trigger a regeneration when either an endpoint or a model file is updated. This will be a necessary change once we listen to all files in the project directory to prevent unnecessary regenerations for other file changes.

A small delay was added to generate files so that multiple changes are batched together. For example, when a file is moved from one folder to another, this creates an `add` and a `remove` event, which could introduce an invalid state if they are processed separately. We already have tests that validate this in the e2e generate watch tests.

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

None - the changes in this PR should preserve the current generated behavior.

### CODE REVIEW

File: tools/serverpod_cli/lib/src/commands/generate.dart [hunk 1]
```diff
@@ -59,18 +61,25 @@ class GenerateCommand extends ServerpodCommand {
         Directory(path.joinAll(config.endpointsSourcePathParts));
     var endpointsAnalyzer = EndpointsAnalyzer(endpointDirectory);
 
+    var protocols = await ModelHelper.loadProjectYamlModelsFromDisk(config);
```
Comment (Isakdl) [inline tools/serverpod_cli/lib/src/commands/generate.dart:64]: Suggestion: Rename protocols to models, as that is what we are loading.

Comment (Isakdl) [reply tools/serverpod_cli/lib/src/commands/generate.dart:64]: Or perhaps yamlModels is more accurate

File: tools/serverpod_cli/lib/src/generator/generator_continuous.dart [hunk 2]
```diff
@@ -32,6 +33,8 @@ Future<bool> performGenerateContinuously({
 
   var modelSourcePath = p.joinAll(config.modelSourcePathParts);
   var protocolSourcePath = p.joinAll(config.protocolSourcePathParts);
+
+  Timer? generateDispatch;
```
Comment (Isakdl) [inline tools/serverpod_cli/lib/src/generator/generator_continuous.dart:37]: Suggestion(nit): Rename to debounceDispatch as that is what we are doing!

**Review by Isakdl — APPROVED** [review]
LGTM, but I have added some nit suggestions!

File: tools/serverpod_cli/lib/src/analyzer/dart/endpoints_analyzer.dart [hunk 3]
```diff
@@ -176,6 +201,17 @@ class EndpointsAnalyzer {
     }
   }
 
+  bool _isEndpointFile(File file) {
+    if (!file.absolute.path.startsWith(absoluteIncludedPaths)) return false;
+    if (!file.path.endsWith('.dart')) return false;
+    if (!file.existsSync()) return false;
+
+    var contents = file.readAsStringSync();
+    if (!contents.contains('extends Endpoint')) return false;
```
[the remark this answers is not in the document]
Comment (SandPod) [reply tools/serverpod_cli/lib/src/analyzer/dart/endpoints_analyzer.dart:210]: Very true! This should work as our source of truth around what files we analyze. So we could update this spot when that happens.

**Pull request merged**